idlc does not have an option to respect directory structure (feature?)

I'm working on a project that makes use of a large collection of generated IDL files. 
The structure of the files is fairly complex with 4-5 layers of sub-directories. Many of the IDL files have `#include` headers for dependencies to other structs, modules, etc. within this structure
```
Base
---- Sub1
-------- Sub1A
------------ Sub1AA
---------------- Sub1AAA.idl
------------ Sub1AB
---------------- Sub1ABA.idl
------------ Sub1AC
---------------- Sub1ACA.idl
-------- Sub1B
-------- Sub1C
-------- ....
---- Sub2
---- Sub3
---- Sub4
---- ....
```
When using `idlcxx_generate` with cmake, all the files are just thrown into the cmake bin directory and don't retain any of the directory information, **but** the `include` directives are preserved which leads to compilation errors.

## Example

**TL;DR** You have to manually recreate the directory structure or use an external script, I want a flag for idlc to create a directory structure matching the path from a user-defined root. I'll do the PR myself if it's something you're open to

I have this directory structure with IDL files
![image](https://user-images.githubusercontent.com/84148896/159345671-92c49ae8-1643-4af0-8861-64d826f7f028.png)
### IDL code
https://gist.github.com/ahoenerBE/812a9baaeef2c63c652357bb728f59da

### idlcxx_generate output
```
file(GLOB_RECURSE idl_files RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} idl/**/*.idl)

idlcxx_generate(TARGET test_idls
        FILES ${idl_files}
        INCLUDES ${CMAKE_CURRENT_SOURCE_DIR}/idl
        FEATURES case-sensitive
        DEFAULT_EXTENSIBILITY appendable
        WARNINGS no-implicit-extensibility
        )
```
![image](https://user-images.githubusercontent.com/84148896/159346520-78782bc1-b2a8-41a5-8eb2-11b11f874bfb.png)

Inside the files, the idl `#includes` have been translated correctly to the right output headers, but they reference directories that don't exist
![image](https://user-images.githubusercontent.com/84148896/159346893-1cf4d617-0767-46a9-8101-c53c8c29af95.png)
```
FAILED: CMakeFiles/ddscxxHelloworldPublisher.dir/Item.cpp.o 
/usr/bin/c++  -I/home/drew/CLionProjects/CycloneTests/cmake-build-debug -isystem /home/drew/CLionProjects/cyclonedds-cxx/install/include/ddscxx -isystem /home/drew/CLionProjects/cyclonedds/install/include -g -std=gnu++17 -MD -MT CMakeFiles/ddscxxHelloworldPublisher.dir/Item.cpp.o -MF CMakeFiles/ddscxxHelloworldPublisher.dir/Item.cpp.o.d -o CMakeFiles/ddscxxHelloworldPublisher.dir/Item.cpp.o -c /home/drew/CLionProjects/CycloneTests/cmake-build-debug/Item.cpp
In file included from /home/drew/CLionProjects/CycloneTests/cmake-build-debug/Item.cpp:9:
/home/drew/CLionProjects/CycloneTests/cmake-build-debug/Item.hpp:12:10: fatal error: Base/Inventory/ItemCategory.hpp: No such file or directory
   12 | #include "Base/Inventory/ItemCategory.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

Either using a script or manually recreating the directory structure allows the program to compile
![image](https://user-images.githubusercontent.com/84148896/159347646-0ed0304b-b03d-473a-9273-c510034233d0.png)

## Request
Could there be an idlc flag to recreate a directory structure starting at a certain root, like the include path flag?
If so, would you be open to a PR for it, I'd like to have it integrated in the tool if I could

--- a/cmake/Modules/Generate.cmake
+++ b/cmake/Modules/Generate.cmake
@@ -12,7 +12,7 @@
 
 function(IDLC_GENERATE)
   set(options NO_TYPE_INFO)
-  set(one_value_keywords TARGET DEFAULT_EXTENSIBILITY)
+  set(one_value_keywords TARGET DEFAULT_EXTENSIBILITY BASE_DIR)
   set(multi_value_keywords FILES FEATURES INCLUDES WARNINGS)
   cmake_parse_arguments(
     IDLC "${options}" "${one_value_keywords}" "${multi_value_keywords}" "" ${ARGN})
@@ -20,6 +20,7 @@ function(IDLC_GENERATE)
   set(gen_args
     ${IDLC_UNPARSED_ARGUMENTS}
     TARGET ${IDLC_TARGET}
+    BASE_DIR ${IDLC_BASE_DIR}
     FILES ${IDLC_FILES}
     FEATURES ${IDLC_FEATURES}
     INCLUDES ${IDLC_INCLUDES}
@@ -33,7 +34,7 @@ endfunction()
 
 function(IDLC_GENERATE_GENERIC)
   set(options NO_TYPE_INFO)
-  set(one_value_keywords TARGET BACKEND DEFAULT_EXTENSIBILITY)
+  set(one_value_keywords TARGET BACKEND DEFAULT_EXTENSIBILITY BASE_DIR)
   set(multi_value_keywords FILES FEATURES INCLUDES WARNINGS SUFFIXES DEPENDS)
   cmake_parse_arguments(
     IDLC "${options}" "${one_value_keywords}" "${multi_value_keywords}" "" ${ARGN})
@@ -113,7 +114,13 @@ function(IDLC_GENERATE_GENERIC)
     list(APPEND IDLC_ARGS "-t")
   endif()
 
+  if(IDLC_BASE_DIR)
+    file(REAL_PATH ${IDLC_BASE_DIR} _base_dir_abs)
+    list(APPEND IDLC_ARGS "-b${_base_dir_abs}")
+  endif()
+
   set(_dir ${CMAKE_CURRENT_BINARY_DIR})
+  list(APPEND IDLC_ARGS "-o${_dir}")
   set(_target ${IDLC_TARGET})
   foreach(_file ${IDLC_FILES})
     get_filename_component(_path ${_file} ABSOLUTE)
@@ -127,11 +134,29 @@ function(IDLC_GENERATE_GENERIC)
   set(_outputs "")
   foreach(_file ${_files})
     get_filename_component(_name ${_file} NAME_WE)
+    get_filename_component(_name_ext ${_file} NAME)
+
+    # Determine middle path for directory reconstruction
+    if(IDLC_BASE_DIR)
+      file(RELATIVE_PATH _file_path_rel ${_base_dir_abs} ${_file})
+      # Hard Fail
+      if(_file_path_rel MATCHES "^\\.\\.")
+        message(FATAL_ERROR "Cannot use base dir with different file tree from input file (${_base_dir_abs} to ${_file} yields ${_file_path_rel})")
+      endif()
+      string(REPLACE ${_name_ext} "" _mid_dir_path ${_file_path_rel})
+      string(REGEX REPLACE "[\\/]$" "" _mid_dir_path ${_mid_dir_path})
+    endif()
+
     set(_file_outputs "")
     foreach(_suffix ${IDLC_SUFFIXES})
-      list(APPEND _file_outputs "${_dir}/${_name}${_suffix}")
-      list(APPEND _outputs ${_file_outputs})
+      if(IDLC_BASE_DIR)
+        list(APPEND _file_outputs "${_dir}/${_mid_dir_path}/${_name}${_suffix}")
+      else()
+        list(APPEND _file_outputs "${_dir}/${_name}${_suffix}")
+      endif()
     endforeach()
+
+    list(APPEND _outputs ${_file_outputs})
     add_custom_command(
       OUTPUT   ${_file_outputs}
       COMMAND  ${_idlc_executable}

--- a/src/idl/include/idl/file.h
+++ b/src/idl/include/idl/file.h
@@ -15,6 +15,9 @@
 #include "idl/export.h"
 #include "idl/retcode.h"
 
+#define IDL_GEN_OUTFILE(path, output_dir, base_dir, out_ext, out_ptr) \
+  idl_generate_out_file(path, output_dir, base_dir, out_est, out_ptr, false)
+
 IDL_EXPORT unsigned int
 idl_isseparator(int chr);
 
@@ -32,5 +35,8 @@ idl_relative_path(const char *base, const char *path, char **relpathp);
 
 IDL_EXPORT idl_retcode_t
 idl_mkpath(const char *path);
+
+IDL_EXPORT idl_retcode_t
+idl_generate_out_file(const char *path, const char *output_dir, const char *base_dir, const char *out_ext, char ** out_ptr, int skip_mkpath);
 
 #endif /* IDL_FILE_H */

--- a/src/idl/include/idl/file.h
+++ b/src/idl/include/idl/file.h
@@ -30,4 +30,7 @@ idl_normalize_path(const char *path, char **abspathp);
 IDL_EXPORT idl_retcode_t
 idl_relative_path(const char *base, const char *path, char **relpathp);
 
+IDL_EXPORT idl_retcode_t
+idl_mkpath(const char *path);
+
 #endif /* IDL_FILE_H */

--- a/src/idl/src/file.c
+++ b/src/idl/src/file.c
@@ -414,3 +414,158 @@ idl_retcode_t idl_mkpath(const char *path)
   err_full_path:
   return ret;
 }
+
+/**
+ * \verbatim
+ *  Given inputs path, output_dir, base_dir, out_ext, constructs the correct output path for the file.
+ *  Unless specified otherwise, will also create intermediate directories to allow for immediate access
+ *  to the file on return.
+ *
+ *  If the input file is absolute, the output path will always be in the form of
+ *  cwd/file.ext OR output_dir/file.ext
+ *  File is derived from the end of `path`, any directories `a, b, c, etc.` are derived from rel. path difference
+ *  between `path` and `base_dir`.
+ *
+ *  If the input path is relative, output path will be in the form of
+ *  a/b/c/file.ext OR output_dir/a/b/c/file.ext, given an input path like a/b/c/file.idl
+ *
+ *  If base_dir is provided, output_path will take the form of
+ *  b/c/file.ext OR output_dir/b/c/file.ext given
+ *  input path /a/b/c/file.idl and a base path /a
+ *  Memory will be allocated to ensure `path` is absolute, to use idl_relative_path and determine the common path.
+ *  This will return an error if the path to the base_dir requires us to go up the tree,
+ *  which could lead to unwanted behavior.
+ *  (e.g. path=/a/b/c/d/file.idl, base_path=/a/b/c/f => ../d/file.idl)
+ * \endverbatim
+ *
+ * @param path          Path (abs / rel) to the input file. Non NULL
+ * @param output_dir    Path (abs / rel) to the desired output directory. Can be NULL
+ * @param base_dir      Absolute path to the "base directory" of a tree. Can be NULL
+ * @param out_ext       Desired output extension. Can be NULL
+ * @param out_ptr       Pointer to output str. Non NULL
+ * @param skip_mkpath   Allows the function to skip making the output path. Used for testing, recommended False
+ * @return retcode
+ */
+idl_retcode_t idl_generate_out_file(const char *path, const char *output_dir, const char *base_dir, const char *out_ext, char ** out_ptr, int skip_mkpath) {
+  assert(out_ptr);
+
+  idl_retcode_t ret = IDL_RETCODE_NO_MEMORY;
+  const char *sepr, *ext, *file;
+  char empty[1] = { '\0' };
+  char *dir = NULL, *basename = NULL, *abs_file_path = NULL;
+  char* rel_path = NULL;
+  char *output_path = NULL;
+
+  if(base_dir && !idl_isabsolute(base_dir)) {
+    ret = IDL_RETCODE_BAD_PARAMETER;
+    goto err_dir;
+  }
+
+  sepr = ext = NULL;
+  for (const char *ptr = path; ptr[0]; ptr++) {
+    if (idl_isseparator((unsigned char)ptr[0]) && ptr[1] != '\0')
+      sepr = ptr;
+    else if (ptr[0] == '.')
+      ext = ptr;
+  }
+
+  file = sepr ? sepr + 1 : path;
+  if (idl_isabsolute(path) || !sepr)
+    dir = empty;
+  else if (!(dir = idl_strndup(path, (size_t)(sepr-path))))
+    goto err_dir;
+  if (!(basename = idl_strndup(file, ext ? (size_t)(ext-file) : strlen(file))))
+    goto err_basename;
+
+  /* replace backslashes by forward slashes */
+  for (char *ptr = dir; *ptr; ptr++) {
+    if (*ptr == '\\')
+      *ptr = '/';
+  }
+
+  if(base_dir) {
+    char* old_dir = dir;
+
+    // We need an absolute path to use idl_relative_path so grab the absolute path of our input file
+    // if we were provided relative paths
+    if(!idl_isabsolute(path)) {
+      if((idl_normalize_path(path, &abs_file_path)) < 0) {
+        goto err_rel_path;
+      }
+    } else {
+      if(!(abs_file_path = idl_strdup(path))) {
+        goto err_rel_path;
+      }
+    }
+
+    // If we run into a path error we can still recover and use existing dir
+    if(idl_relative_path(base_dir, abs_file_path, &rel_path) == IDL_RETCODE_NO_MEMORY)
+      goto err_rel_path;
+    // If root is not a parent of file path, or there was an error resolving rel path then return an error
+#if _WIN32
+    int is_reverse_path = (rel_path && (strncmp(rel_path, "..\\", 3) == 0));
+#else
+    int is_reverse_path = (rel_path && (strncmp(rel_path, "../", 3) == 0));
+#endif
+    if(is_reverse_path){
+      ret = IDL_RETCODE_BAD_PARAMETER;
+      goto err_rel_path;
+    }
+
+    // Extract the common directory to dir, free the old_dir pointer
+    dir = NULL;
+    size_t print_len = rel_path == NULL ? 0 : strlen(rel_path);
+    if (file) {
+      size_t file_len = sepr == NULL ? 0 : strlen(sepr);
+      print_len = print_len >= file_len ? print_len - file_len : 0;
+    }
+    if (!(dir = idl_strndup(rel_path, print_len))) {
+      goto err_rel_path;
+    }
+    if (old_dir && old_dir[0] != '\0')
+      free(old_dir);
+  }
+
+  sepr = dir[0] == '\0' ? "" : "/";
+  if(output_dir && output_dir[0] != '\0') {
+    if(idl_asprintf(&output_path, "%s%s%s", output_dir, sepr, dir) < 0)
+      goto err_rel_path;
+    sepr = "/";
+  } else {
+    if(!(output_path = idl_strdup(dir)))
+      goto err_rel_path;
+  }
+
+  // Allow skipping of path generation for unit testing
+  if(!skip_mkpath) {
+    if(idl_mkpath(output_path) < 0)
+      goto err_mkpath;
+  }
+
+  if (idl_asprintf(out_ptr, "%s%s%s%s%s",
+                   output_path,
+                   sepr,
+                   basename,
+                   out_ext ? "." : "",
+                   out_ext ? out_ext : ""
+                   ) < 0) {
+    goto err_outpath;
+  }
+  ret = IDL_RETCODE_OK;
+
+err_outpath:
+err_mkpath:
+  if(output_path)
+    free(output_path);
+err_rel_path:
+  if(rel_path)
+    free(rel_path);
+  if(abs_file_path)
+    free(abs_file_path);
+  free(basename);
+err_basename:
+  if (dir && dir != empty)
+    free(dir);
+err_dir:
+  return ret;
+}

--- a/src/idl/src/file.c
+++ b/src/idl/src/file.c
@@ -363,3 +363,54 @@ idl_retcode_t idl_relative_path(const char *base, const char *path, char **relpa
   *relpathp = rel;
   return IDL_RETCODE_OK;
 }
+
+#if _WIN32
+static inline idl_retcode_t idl_mkdir(const char *pathname, int mode) {
+  (void) mode;
+  return _mkdir(pathname);
+}
+#else
+static inline idl_retcode_t idl_mkdir(const char *pathname, mode_t mode) {
+  return mkdir(pathname, mode);
+}
+#endif
+
+idl_retcode_t idl_mkpath(const char *path)
+{
+  idl_retcode_t ret = IDL_RETCODE_NO_MEMORY;
+  char *full_path;
+
+  assert(path);
+  if (!(full_path = absolute_path(path)))
+    goto err_full_path;
+
+  if(idl_untaint_path(full_path) < 0)
+    goto err_untaint;
+
+  {
+    char chr, *ptr = full_path;
+
+    for (; *ptr && *ptr != '/' && *ptr != sep; ptr++)
+      /* skip ahead to first separator to skip drive */;
+
+    while (*ptr)
+    {
+      for (++ptr; *ptr && *ptr != '/' && *ptr != sep; ptr++)
+        /* search for next segment */;
+      assert(ptr[-1] != '/' && ptr[-1] != sep);
+      chr = *ptr;
+      *ptr = '\0';
+      if ((ret = idl_mkdir(full_path, 0777)) == -1 && errno != EEXIST)
+        goto err_mkdir;
+      *ptr = chr;
+    }
+
+    ret = IDL_RETCODE_OK;
+  }
+
+  err_mkdir:
+  err_untaint:
+  free(full_path);
+  err_full_path:
+  return ret;
+}

--- a/src/idl/tests/file.c
+++ b/src/idl/tests/file.c
@@ -324,3 +324,56 @@ CU_Test(idl_file, relative)
       free(rel);
   }
 }
+
+
+typedef struct out_file_test{
+  const char *path;
+  const char *output_dir;
+  const char *base_dir;
+  const char *out_ext;
+  idl_retcode_t ret;
+  const char *expected_out;
+} out_file_test_t;
+
+static void test_out_file(const out_file_test_t test)
+{
+  char *out = NULL;
+  idl_retcode_t ret = idl_generate_out_file(test.path, test.output_dir, test.base_dir, test.out_ext, &out, true);
+
+  CU_ASSERT_EQUAL(ret, test.ret);
+
+  if(ret != IDL_RETCODE_BAD_PARAMETER){
+    CU_ASSERT_STRING_EQUAL(out, test.expected_out);
+  }
+
+  if (out)
+    free(out);
+}
+
+CU_Test(idl_file, out_file_generation)
+{
+  const out_file_test_t out_tests[] = {
+      {"a/b.idl",            NULL,    NULL,         NULL, IDL_RETCODE_OK,            "a/b"},
+      {"a/b.idl",            NULL,    NULL,         "c",  IDL_RETCODE_OK,            "a/b.c"},
+      {"a/b.idl",            "c",     NULL,         NULL, IDL_RETCODE_OK,            "c/a/b"},
+      {ROOT"a/b.idl",        NULL,    ROOT"a",      "c",  IDL_RETCODE_OK,            "b.c"},
+      {ROOT"a/b.idl",        "f",     ROOT"a",      "c",  IDL_RETCODE_OK,            "f/b.c"},
+      {ROOT"a/b.idl",        NULL,    NULL,         NULL, IDL_RETCODE_OK,            "b"},
+      {ROOT"a/b.idl",        NULL,    NULL,         "c",  IDL_RETCODE_OK,            "b.c"},
+      {"a/b.idl",            "c",     NULL,         NULL, IDL_RETCODE_OK,            "c/a/b"},
+      {ROOT"a/b.idl",        "c",     NULL,         NULL, IDL_RETCODE_OK,            "c/b"},
+      {"a/b.idl",            ROOT"c", NULL,         NULL, IDL_RETCODE_OK,            ROOT"c/a/b"},
+      {"a/b/c.idl",          "..",    NULL,         NULL, IDL_RETCODE_OK,            "../a/b/c"},
+      {"a/b/c.idl",          "../d",  NULL,         NULL, IDL_RETCODE_OK,            "../d/a/b/c"},
+      {"a/b/c.idl",          ".",     NULL,         NULL, IDL_RETCODE_OK,            "./a/b/c"},
+      {ROOT"a/b/c/file.idl", ".",     ROOT"a/b/",   NULL, IDL_RETCODE_OK,            "./c/file"},
+      {ROOT"a/b/c/file.idl", "..",    ROOT"a/b/",   "h",  IDL_RETCODE_OK,            "../c/file.h"},
+      {ROOT"a/b/c/file.idl", ".",     ROOT"a/b/",   NULL, IDL_RETCODE_OK,            "./c/file"},
+      {ROOT"a/b/c/file.idl", NULL,    ROOT"a/b/",   NULL, IDL_RETCODE_OK,            "c/file"},
+      {ROOT"a/b/c/file.idl", ".",     ROOT"a/b/d/", NULL, IDL_RETCODE_BAD_PARAMETER, "./c/file"},
+      {ROOT"a/b/c/file.idl", NULL,    ROOT"a/b/d/", NULL, IDL_RETCODE_BAD_PARAMETER, "c/file"}
+  };
+
+  for (size_t i = 0; i < sizeof(out_tests)/sizeof(out_tests[0]); i++)
+    test_out_file(out_tests[i]);
+}

--- a/src/tools/idlc/include/idlc/generator.h
+++ b/src/tools/idlc/include/idlc/generator.h
@@ -28,6 +28,9 @@ extern "C" {
 
 typedef struct idlc_generator_config idlc_generator_config_t;
 struct idlc_generator_config  {
+  char *output_dir; /* path to write completed files */
+  char* base_dir; /* Path to start reconstruction of dir structure */
+
   /** Flag to indicate if xtypes type information is included in the generated types */
   bool generate_type_info;
 

--- a/src/tools/idlc/src/generator.c
+++ b/src/tools/idlc/src/generator.c
@@ -362,6 +362,8 @@ idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
   const char *sep, *ext, *file, *path;
   char empty[1] = { '\0' };
   char *dir = NULL, *basename = NULL;
+  char* rel_path = NULL;
+  char *output_path = NULL;
   struct generator generator;
 
   assert(pstate->paths);
@@ -369,7 +371,7 @@ idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
   assert(config);
   path = pstate->sources->path->name;
   /* use relative directory if user provided a relative path, use current
-     word directory otherwise */
+     work directory otherwise */
   sep = ext = NULL;
   for (const char *ptr = path; ptr[0]; ptr++) {
     if (idl_isseparator((unsigned char)ptr[0]) && ptr[1] != '\0')
@@ -395,12 +397,43 @@ idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
   memset(&generator, 0, sizeof(generator));
   generator.path = file;
 
+  if(config->base_dir) {
+    char* old_dir = dir;
+    // If we run into a path error we can still recover and use existing dir
+    if(idl_relative_path(config->base_dir, pstate->sources->path->name, &rel_path) == IDL_RETCODE_NO_MEMORY)
+      goto err_rel_path;
+    // If root is not a parent of file path, or there was an error resolving rel path then skip
+    if(rel_path != NULL && strncmp(rel_path, "../", 3) != 0) {
+      dir = NULL;
+      size_t print_len = strlen(rel_path);
+      if (sep)
+        print_len -= strlen(sep);
+      if (!(dir = idl_strndup(rel_path, print_len))) {
+        goto err_rel_path;
+      }
+      if (old_dir && old_dir[0] != '\0')
+        free(old_dir);
+    }
+  }
+
   sep = dir[0] == '\0' ? "" : "/";
-  if (idl_asprintf(&generator.header.path, "%s%s%s.h", dir, sep, basename) < 0)
+  if(config->output_dir && config->output_dir[0] != '\0') {
+    if(idl_asprintf(&output_path, "%s%s%s", config->output_dir, sep, dir) < 0)
+      goto err_rel_path;
+    sep = "/";
+  } else {
+    if(!(output_path = idl_strdup(dir)))
+      goto err_rel_path;
+  }
+
+  if(idl_mkpath(output_path) < 0)
+    goto err_mkpath;
+
+  if (idl_asprintf(&generator.header.path, "%s%s%s.h", output_path, sep, basename) < 0)
     goto err_header;
   if (!(generator.header.handle = idl_fopen(generator.header.path, "wb")))
     goto err_header;
-  if (idl_asprintf(&generator.source.path, "%s%s%s.c", dir, sep, basename) < 0)
+  if (idl_asprintf(&generator.source.path, "%s%s%s.c", output_path, sep, basename) < 0)
     goto err_source;
   if (!(generator.source.handle = idl_fopen(generator.source.path, "wb")))
     goto err_source;
@@ -426,6 +459,12 @@ err_header:
     fclose(generator.header.handle);
   if (generator.header.path)
     free(generator.header.path);
+err_mkpath:
+  if (output_path)
+    free(output_path);
+err_rel_path:
+  if(rel_path)
+    free(rel_path);
   if (basename)
     free(basename);
 err_basename:

--- a/src/tools/idlc/src/generator.c
+++ b/src/tools/idlc/src/generator.c
@@ -359,81 +359,21 @@ idl_retcode_t
 idlc_generate(const idl_pstate_t *pstate, const idlc_generator_config_t *config)
 {
   idl_retcode_t ret = IDL_RETCODE_NO_MEMORY;
-  const char *sep, *ext, *file, *path;
-  char empty[1] = { '\0' };
-  char *dir = NULL, *basename = NULL;
-  char* rel_path = NULL;
-  char *output_path = NULL;
   struct generator generator;
 
   assert(pstate->paths);
   assert(pstate->paths->name);
   assert(config);
-  path = pstate->sources->path->name;
-  /* use relative directory if user provided a relative path, use current
-     work directory otherwise */
-  sep = ext = NULL;
-  for (const char *ptr = path; ptr[0]; ptr++) {
-    if (idl_isseparator((unsigned char)ptr[0]) && ptr[1] != '\0')
-      sep = ptr;
-    else if (ptr[0] == '.')
-      ext = ptr;
-  }
-
-  file = sep ? sep + 1 : path;
-  if (idl_isabsolute(path) || !sep)
-    dir = empty;
-  else if (!(dir = idl_strndup(path, (size_t)(sep-path))))
-    goto err_dir;
-  if (!(basename = idl_strndup(file, ext ? (size_t)(ext-file) : strlen(file))))
-    goto err_basename;
-
-  /* replace backslashes by forward slashes */
-  for (char *ptr = dir; *ptr; ptr++) {
-    if (*ptr == '\\')
-      *ptr = '/';
-  }
+  const char* path = pstate->sources->path->name;
 
   memset(&generator, 0, sizeof(generator));
-  generator.path = file;
+  generator.path = path;
 
-  if(config->base_dir) {
-    char* old_dir = dir;
-    // If we run into a path error we can still recover and use existing dir
-    if(idl_relative_path(config->base_dir, pstate->sources->path->name, &rel_path) == IDL_RETCODE_NO_MEMORY)
-      goto err_rel_path;
-    // If root is not a parent of file path, or there was an error resolving rel path then skip
-    if(rel_path != NULL && strncmp(rel_path, "../", 3) != 0) {
-      dir = NULL;
-      size_t print_len = strlen(rel_path);
-      if (sep)
-        print_len -= strlen(sep);
-      if (!(dir = idl_strndup(rel_path, print_len))) {
-        goto err_rel_path;
-      }
-      if (old_dir && old_dir[0] != '\0')
-        free(old_dir);
-    }
-  }
-
-  sep = dir[0] == '\0' ? "" : "/";
-  if(config->output_dir && config->output_dir[0] != '\0') {
-    if(idl_asprintf(&output_path, "%s%s%s", config->output_dir, sep, dir) < 0)
-      goto err_rel_path;
-    sep = "/";
-  } else {
-    if(!(output_path = idl_strdup(dir)))
-      goto err_rel_path;
-  }
-
-  if(idl_mkpath(output_path) < 0)
-    goto err_mkpath;
-
-  if (idl_asprintf(&generator.header.path, "%s%s%s.h", output_path, sep, basename) < 0)
+  if (idl_generate_out_file(path, config->output_dir, config->base_dir, "h", &generator.header.path, false) < 0)
     goto err_header;
   if (!(generator.header.handle = idl_fopen(generator.header.path, "wb")))
     goto err_header;
-  if (idl_asprintf(&generator.source.path, "%s%s%s.c", output_path, sep, basename) < 0)
+  if (idl_generate_out_file(path, config->output_dir, config->base_dir, "c", &generator.source.path, false) < 0)
     goto err_source;
   if (!(generator.source.handle = idl_fopen(generator.source.path, "wb")))
     goto err_source;
@@ -459,17 +399,5 @@ err_header:
     fclose(generator.header.handle);
   if (generator.header.path)
     free(generator.header.path);
-err_mkpath:
-  if (output_path)
-    free(output_path);
-err_rel_path:
-  if(rel_path)
-    free(rel_path);
-  if (basename)
-    free(basename);
-err_basename:
-  if (dir && dir != empty)
-    free(dir);
-err_dir:
   return ret;
 }

--- a/src/tools/idlc/src/idlc.c
+++ b/src/tools/idlc/src/idlc.c
@@ -24,7 +24,6 @@
 #include "idl/tree.h"
 #include "idl/string.h"
 #include "idl/processor.h"
-#include "idl/file.h"
 #include "idl/version.h"
 #include "idl/stream.h"
 
@@ -36,6 +35,7 @@
 #include "plugin.h"
 #include "options.h"
 #include "descriptor_type_meta.h"
+#include "file.h"
 
 #if 0
 #define IDLC_DEBUG_PREPROCESSOR (1u<<2)
@@ -54,6 +54,8 @@ struct idlc_disable_warning_list
 static struct {
   char *file; /* path of input file or "-" for STDIN */
   const char *lang;
+  const char *output_dir; /* path to write completed files */
+  const char* base_dir; /* Path to start reconstruction of dir structure */
   int compile;
   int preprocess;
   int keylist;
@@ -544,6 +546,15 @@ static const idlc_option_t *compopts[] = {
     "Enable or disable warnings. Possible values are: no-implicit-extensibility, "
     "no-extra-token-directive, no-unknown_escape_seq, no-inherit-appendable, "
     "no-eof-newline, no-enum-consecutive. " },
+  &(idlc_option_t){
+    IDLC_STRING, { .string = &config.output_dir }, 'o', "", "<directory>",
+    "Set the output directory for compiled files." },
+  &(idlc_option_t){
+    IDLC_STRING, { .string = &config.base_dir }, 'b', "", "<directory>",
+    "Enable directory reconstruction starting from <base_dir> "
+    "Attempts to recreate directory structure matching input. "
+    "Use without the -o option will compile IDL files in-place. "
+    "Fail-silent if root_dir is not a parent of input" },
 #ifdef DDS_HAS_TYPE_DISCOVERY
   &(idlc_option_t){
     IDLC_FLAG, { .flag = &config.no_type_info }, 't', "", "",
@@ -676,6 +687,19 @@ int main(int argc, char *argv[])
     } else if (config.compile) {
       idlc_generator_config_t generator_config;
       memset(&generator_config, 0, sizeof(generator_config));
+
+      // Duplicate/Untaint the output dir to keep header guards neat
+      if(config.output_dir) {
+        if(!(generator_config.output_dir = idl_strdup(config.output_dir)))
+          goto err_generate;
+        if(idl_untaint_path(generator_config.output_dir) < 0)
+          goto err_generate;
+      }
+      // Root dir must be normalized because relativity comparison will be done
+      if(config.base_dir) {
+        if(idl_normalize_path(config.base_dir, &generator_config.base_dir) < 0)
+          goto err_generate;
+      }
 #ifdef DDS_HAS_TYPE_DISCOVERY
       if(!config.no_type_info)
         generator_config.generate_type_info = true;
@@ -683,6 +707,11 @@ int main(int argc, char *argv[])
 #endif // DDS_HAS_TYPE_DISCOVERY
       if (gen.generate)
         ret = gen.generate(pstate, &generator_config);
+
+      if(generator_config.output_dir)
+        free(generator_config.output_dir);
+      if(generator_config.base_dir)
+        free(generator_config.base_dir);
       idl_delete_pstate(pstate);
       if (ret) {
         fprintf(stderr, "Failed to compile '%s'\n", config.file);

